### PR TITLE
Improve concurrency

### PR DESCRIPTION
**Pull request by iProdigy**
Use one ScheduledThreadPoolExecutor for broadcasting remaining requests to connected web sockets instead of creating a new Timer for each socket (which would pollute the environment with many threads). Use this timer to schedule pings to github to verify the number of remaining requests.

### CODE REVIEW

File: src/main/kotlin/app/Main.kt [hunk 1]
```diff
@@ -61,9 +65,23 @@ fun main(args: Array<String>) {
             }
         }
 
+        // Keep track of the rate limit broadcast actions associated with each socket to easily cancel on close
```
Comment (iProdigy) [inline src/main/kotlin/app/Main.kt]: Should this be moved elsewhere (e.g. a socket broadcast manager)?

Comment (tipsy) [reply src/main/kotlin/app/Main.kt]: Yes, sounds a lot cleaner

**Comment by tipsy** [comment]
Would be easier to review if you could split this into different PRs.

**Comment by iProdigy** [comment]
@tipsy agree that it was a bit unfocused - what else should be moved to a separate PR

**Comment by tipsy** [comment]
I think the changes here all can go in the same PR, but I'm not sure about the code organization. Maybe both timers should go in a new `GitHubRateLimitUtil`, which is both responsible for pinging GH and broadcasting to clients 🤔

I'll look at it more closely tomorrow.

**Comment by tipsy** [comment]
If you're keeping track of the websocket-sessions, you don't really need one executor for each, you can make one executor that loops through them all:

```

// GhService:

...all the client/service things
  
fun pingGhClients() { ... }

val wsClients = ConcurrentLinkedQueue<WsSession>
fun pingWsClients() { ... }    

init {
    Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate({
        GhService.pingGhClients()
    }, 0, 2, TimeUnit.MINUTES)
    Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate({
        GhService.pingWsClients()
    }, 0, 500, TimeUnit.MILLISECONDS)
}
        


// Main:
ws("/rate-limit-status") { ws ->
    ws.onConnect { session -> GhService.wsClients.add(session) }
    ws.onClose { session, _, _ -> GhService.wsClients.remove(session) }
    ws.onError { session, _ -> GhService.wsClients.remove(session) }
}
```

**Comment by tipsy** [comment]
You can also pull out the `RateLimitUtil` changes to a separate PR.

**Comment by iProdigy** [comment]
@tipsy Any changes to be made?

**Comment by tipsy** [comment]
@iProdigy I'll have a more thorough look after work, but a few things:

* Why a map?
* Why register/unregister instead of just using the map directly
* `it.send(payload)` should probably be in a `try`

**Comment by tipsy** [comment]
Didn't really see anything else on going through it again.

**Comment by iProdigy** [comment]
@tipsy,

I went for a map because I felt it better fit the situation; there is frequent iteration and add+remove operations while the hashcode function seems to be of sufficient quality.
* Like a queue, adding/offering an element is O(1).
* With linked queues, iteration is sequential. A ConcurrentHashMap, on the other hand, allows for parallel iteration of its elements which is preferable when you have many clients that should be receiving updates at the same time. 
* With linked queues, removing an element is O(n) since the queue needs to sequentially walk to the element to be removed. Hash maps, on the other hand, remove elements in constant time (on average). This by itself is a large advantage, especially since removals can be concurrent and frequent.

**Comment by iProdigy** [comment]
I used register/unregister functions to better fit an object-oriented paradigm where implementation is abstracted away while outside classes interact with a more generic interface. As a result, the specifics of the implementation can change without requiring change from the classes that may use those functions.

**Comment by tipsy** [comment]
Fair enough. Seems a bit overkill to me, but I'll merge it anyway. Thanks!

**Pull request merged**